gems/sorbet-runtime/lib/types/non_forcing_constants.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+# typed: strict
+
+module T::NonForcingConstants
+  # NOTE: This method is documented on the RBI in Sorbet's payload, so that it
+  # shows up in the hover/completion documentation via LSP.
+  T::Sig::WithoutRuntime.sig {params(val: BasicObject, klass: String).returns(T::Boolean)}
+  def self.non_forcing_is_a?(val, klass)
+    method_name = "T::NonForcingConstants.non_forcing_is_a?"
+    if klass.empty?
+      raise ArgumentError.new("The string given to `#{method_name}` must not be empty")
+    end
+
+    current_klass = T.let(nil, T.nilable(Module))
+    current_prefix = T.let(nil, T.nilable(String))
+
+    parts = klass.split('::')
+    parts.each do |part|
+      if current_klass.nil?
+        # First iteration
+        if part != ""
+          raise ArgumentError.new("The string given to `#{method_name}` must be an absolute constant reference that starts with `::`")
+        end
+
+        current_klass = Object
+        current_prefix = ''
+      else
+        if current_klass.autoload?(part)
+          # There's an autoload registered for that constant, which means it's not
+          # yet loaded. `value` can't be an instance of something not yet loaded.
+          return false
+        end
```

Comment:
I'm a little concerned that the static and runtime semantics differ here. In the static pass, a return value of `false` here means that `val` is definitely not a subclass of `klass`, however here it means that it's either not a subclass, or that `klass` isn't yet loaded. Do you think it's worth returning a third value to indicate the case when `klass` isn't loaded?

Reply:
That's the point: if the class isn't loaded, then it can't be the case that `x.is_a?(ClassNotYetLoaded)` because to have created an instance of it, it would have had to load.

The concerning difference to me between the static and runtime is that the runtime will say `false` where an eager `is_a?` check would have raised a NameError exception if the fact that a constant doesn't exist can only be known by forcing other constants. That's a tradeoff that I'm ok with because the static checks guarantee that the static string resolves to a constant (though you can still lie to it with an RBI for example).

Reply:
Good point, thanks!